contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+                (first_hop.from_token.address, first_hop.from_token.code_hash)
+            {
+                // first hop is a snip20
+                msgs.push(snip20::send_msg(
+                    first_hop.pair_address,
+                    amount,
+                    // build swap msg for the next hop
+                    Some(to_binary(&Snip20Swap::Swap {
+                        // set expected_return to None because we don't care about slippage mid-route
+                        expected_return: None,
+                        // set the recepient of the swap to be this contract (the router)
+                        to: Some(env.contract.address.clone()),
+                    })?),
+                    None,
+                    256,
+                    token_code_hash,
+                    token_address,
+                )?);
+            } else if first_hop.from_token.native_denom == Some("uscrt".into()) {
+                // first hop is SCRT
+                msgs.push(
+                        // build swap msg for the next hop                    CosmosMsg::Wasm(WasmMsg::Execute {
+                        contract_addr: first_hop.pair_address,
+                        callback_code_hash: first_hop.pair_code_hash,
+                        msg: to_binary(&NativeSwap::Swap {
+                            offer_asset: Asset {
+                                amount,
+                                info: AssetInfo::NativeToken {
+                                    denom: "uscrt".into(),
+                                },
+                            },
+                            // set expected_return to None because we don't care about slippage mid-route
+                            expected_return: None,
+                            // set the recepient of the swap to be this contract (the router)
+                            to: Some(env.contract.address.clone()),
+                        })?,
+                        send: vec![Coin::new(amount.u128(), "uscrt")],
+                    }),
+                );
+            } else {
+                // shouldn't be here because we've tested this a few lines before
+                return Err(StdError::generic_err(
+                    "cannot build swap message for first hop",
+                ));
+            }
+
+            msgs.push(
+                // finalize the route at the end, to make sure the route was completed successfully
+                CosmosMsg::Wasm(WasmMsg::Execute {
+                    contract_addr: env.contract.address.clone(),
+                    callback_code_hash: env.contract_code_hash.clone(),
+                    msg: to_binary(&HandleMsg::FinalizeRoute {})?,
+                    send: vec![],
+                }),
+            );
+
+            Ok(HandleResponse {
+                messages: msgs,
+                log: vec![],
+                data: None,
+            })
+        }
+        HandleMsg::Receive {
+            from: _,
```

Comment:
Should probably add a check that `from` is this contract?

Reply:
From is the pair of the previous hop. I can add a check for this, but not sure if it's needed.

Reply:
Sorry, you're correct. Will do.